lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    ch = {
+      contextPreviewExists: false
+
+      removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while editor.session.getContext(i - 1) == "equation"
+          i -= 1
+        start = i
+        while editor.session.getContext(i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
```

Comment:
I see. Yes, if it runs in the event handler then chance are that`this` is going to be different. However, you can pass a closure as the callback, and call `ch`'s methods from the closure, and in this case everything must be fine with `this`